security/advancedtls/crl_provider.go

```diff
@@ -0,0 +1,229 @@
+// FileWatcherCRLProvider. It uses Options to validate and apply configuration
+// required for creating a new instance.
+func MakeFileWatcherCRLProvider(o Options) (*FileWatcherCRLProvider, error) {
+	if err := o.validate(); err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	provider := &FileWatcherCRLProvider{
+		crls: make(map[string]*CRL),
+		opts: o,
+	}
+	provider.cancel = cancel
+	go provider.run(ctx)
+	return provider, nil
+}
+
+func (o *Options) validate() error {
+	// Checks relates to CRLDirectory.
+	if o.CRLDirectory == "" {
+		return fmt.Errorf("advancedtls: CRLDirectory needs to be specified")
+	}
+	fileInfo, err := os.Stat(o.CRLDirectory)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("advancedtls: CRLDirectory %v does not exist", o.CRLDirectory)
+		}
+		return err
+	}
+	if !fileInfo.IsDir() {
+		return fmt.Errorf("advancedtls: CRLDirectory %v is not a directory", o.CRLDirectory)
+	}
+	_, err = os.Open(o.CRLDirectory)
+	if err != nil {
+		if os.IsPermission(err) {
+			return fmt.Errorf("advancedtls: CRLDirectory %v is not readable", o.CRLDirectory)
+		}
+		return err
+	}
+	// Checks related to RefreshDuration.
+	if o.RefreshDuration <= 0 || o.RefreshDuration < time.Second {
+		o.RefreshDuration = defaultCRLRefreshDuration
+		grpclogLogger.Warningf("RefreshDuration must larger then 1 second: provided value %v, default value will be used %v", o.RefreshDuration, defaultCRLRefreshDuration)
+	}
+	return nil
+}
+
+// Start starts watching the directory for CRL files and updates the provider accordingly.
+func (p *FileWatcherCRLProvider) run(ctx context.Context) {
+	ticker := time.NewTicker(p.opts.RefreshDuration)
+	defer ticker.Stop()
+	p.ScanCRLDirectory()
+
+	for {
+		select {
+		case <-ctx.Done():
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			p.ScanCRLDirectory()
+		}
+	}
+}
+
+// Close stops the background refresh of CRLDirectory of FileWatcherCRLProvider
+func (p *FileWatcherCRLProvider) Close() {
+	p.cancel()
+}
+
+// ScanCRLDirectory starts the process of scanning Options.CRLDirectory and
+// updating in-memory storage of CRL structs.Please note that the same method is
+// called periodically by run goroutine.
+func (p *FileWatcherCRLProvider) ScanCRLDirectory() {
+	dir, err := os.Open(p.opts.CRLDirectory)
+	if err != nil {
+		grpclogLogger.Errorf("Can't open CRLDirectory %v", p.opts.CRLDirectory, err)
+		if p.opts.cRLReloadingFailedCallback != nil {
+			p.opts.cRLReloadingFailedCallback(err)
+		}
+	}
+	defer dir.Close()
+
+	files, err := dir.ReadDir(0)
+	if err != nil {
+		grpclogLogger.Errorf("Can't access files under CRLDirectory %v", p.opts.CRLDirectory, err)
+		if p.opts.cRLReloadingFailedCallback != nil {
+			p.opts.cRLReloadingFailedCallback(err)
+		}
+	}
+
+	successCounter := 0
+	failCounter := 0
+	for _, file := range files {
+		filePath := fmt.Sprintf("%s/%s", p.opts.CRLDirectory, file.Name())
+		err := p.addCRL(filePath)
```

Comment:
I think we need some thread safety around these updates, or we need to use a backing map type that is thread safe

Reply:
Done